Gizmo user manual (WIP)

dmlloyd opened the pull request:
Add a user manual for Gizmo 2.x.

Ladicek commented on MANUAL.adoc:
```diff
@@ -59,18 +59,121 @@ The `ClassOutput.fileWriter(Path)` static factory method returns a `ClassWriter`
 [id=classdesc]
 === The `ClassDesc` interface
 
+Types (classes and interfaces, and their variations) are identified using the `ClassDesc` interface, which is provided by the JDK as part of the `java.lang.constant` package.
```
```suggestion
Types (primitives, classes and interfaces, and their variations) are identified using the `ClassDesc` interface, which is provided by the JDK as part of the `java.lang.constant` package.
```

Ladicek commented:
Added a few comments for _Descriptors_, other chapters look fine to me.

mkouba commented on MANUAL.adoc:
```diff
@@ -0,0 +1,419 @@
+public class HelloWorld {
+    public static void main(String[] args) {
+        Gizmo gizmo = Gizmo.create(ClassOutput.fileWriter(Path.of("output")));
+        gizmo.class_("HelloWorldGenerated", zc -> {
```
Why `zc`?

Ladicek replied:
I tend to use `cc` myself, but then when I generate a constructor, I cannot use `cc` again, so I use `mc` even in constructors...

dmlloyd replied:
Yeah I use `cc` for constructors, so then I use `zc` as a mnemonic abbreviation for `clazz creator`, which is a traditional variable name for `Class`-typed things.

Another option is to use `mc` for methods *and* constructors by convention, freeing `cc` for classes.

dmlloyd commented on MANUAL.adoc:
```diff
@@ -0,0 +1,419 @@
+."Hello world" example
+[source,java]
+----
+public class HelloWorld {
```
You're right; I did start doing that later on.

Ladicek commented on MANUAL.adoc:
```diff
@@ -190,6 +190,87 @@ MethodTypeDesc mainType = MethodTypeDesc.of(
+public class SimpleInterfaceExample {
+    public static void main(String[] args) {
+        Gizmo gizmo = Gizmo.create(ClassOutput.fileWriter(Path.of("output")));
+        // create a new class using the given lambda
```
```suggestion
        // create a new interface using the given lambda
```

Ladicek commented on MANUAL.adoc:
```diff
@@ -274,6 +274,79 @@ public class SimpleInterfaceExample {
+Fields may be initialized directly with a constant value at the time the field is defined by using the `setInitial` method.
+For more complex initializers, the `setInitializer` method may be used, which accepts a <<blocks,code block creator>> which must yield a value for the field.
+Initializers are automatically emitted into constructors or static initializers in the order that they are established.
+
```
Since the "Creating types" chapter says

> By convention, the parameter name `cc` is typically used for the `ClassCreator` or `InterfaceCreator` instance.

maybe we should say the similar thing here for field creators?

> By convention, the parameter name `fc` is typically used for the `FieldCreator` instance.

dmlloyd commented:
I've added `Blocks` and `Scoping` today.

Ladicek commented on MANUAL.adoc:
```diff
@@ -806,6 +806,40 @@ public class NaturalOrderingExample {
 
+Every expression has a _type_.
+The type of an expression is generally represented by a <<classdesc,`ClassDesc` descriptor>>.
+The type may be primitive such as `int` or `boolean`, an object type like `java.lang.Object`, or `void`.
```
Also array types.

Ladicek commented on MANUAL.adoc:
```diff
@@ -806,6 +806,40 @@ public class NaturalOrderingExample {
+
+It is sometimes useful to categorize types as either being primitive or non-primitive.
+This is done using the `typeKind()` method, which is also found on `SimpleTyped` and its subtypes, including `Expr`.
+The return type of this method is `TypeKind`, an enumeration which contains constants for each of the eight primitive types, one for object types, and one for `void`.
```
```suggestion
The return type of this method is `TypeKind`, an enumeration which contains constants for each of the eight primitive types, one for reference types, and one for `void`.
```

Ladicek commented on MANUAL.adoc:
```diff
@@ -849,6 +849,27 @@ Most constants can be created once and reused as many times as needed.
+This interface also contains factory methods to create new constants, including (but not limited to):
+
+* Primitive values (e.g. integers, booleans)
+* Primitive wrapper type instances (e.g. `Long`, `Double`)
```
I don't think we have that? The factory methods take wrapper types (such as `Boolean`), but always produce primitive constants (such as `boolean`).

Ladicek commented on MANUAL.adoc:
```diff
@@ -849,6 +849,27 @@ Most constants can be created once and reused as many times as needed.
+* Lists of other object-typed constants (as an immutable `List`)
+* Maps of other object-typed constants (as an immutable `Map`)
+* Other <<jdk-const,JDK constant types>> such as method handles or dynamic constants
+
```
Maybe also mention `Const.ofDefault()` below this list, as a way to obtain a default value of given type (zero for primitives, `null` for references)?

dmlloyd commented:
I pushed up fixes to the code examples so that they all have `main` methods and, when possible, also create classes with `main` methods. This makes it a little easier to test.

I've also added the following sections:

* Yielding values from blocks
* Arithmetic expressions
* Bitwise logical expressions
* Bitwise shift expressions
* Equality and ordering operators
* Comparator-like operations
* Invocation expressions

Ladicek commented on MANUAL.adoc:
```diff
@@ -1657,7 +1657,198 @@ public class NegDoubleExample {
+==== `xor`
+
+The `xor(...)` operation performs a bitwise-XOR on the two arguments, returning the result.
+The Java language equivalent of this operation is the `|` operator.
```
```suggestion
The Java language equivalent of this operation is the `^` operator.
```

Ladicek commented:
Additions LGTM, just a small note: I'm not sure about the claims of commutativity of arithmetic operations in case of floating points. It also feels like the comments about order of operands being [in]significant are superfluous; generally, order of operands is always significant. I'd just drop those lines, but leaving the decision up to you.

Ladicek commented on MANUAL.adoc:
```diff
@@ -2693,7 +2693,58 @@ public class GotoExample {
+                        b1.printf("Both arguments have the same length%n");
+                    });
+                    b0.ifElse(b0.lt(args.elem(0), args.elem(1)), b1 -> {
+                        b1.printf("Argument 0 comes before argument 1");
```
```suggestion
                        b1.printf("Argument 0 comes before argument 1%n");
```

Ladicek commented on MANUAL.adoc:
```diff
@@ -2693,7 +2693,58 @@ public class GotoExample {
+                    b0.ifElse(b0.lt(args.elem(0), args.elem(1)), b1 -> {
+                        b1.printf("Argument 0 comes before argument 1");
+                    }, b1 -> {
+                        b1.printf("Argument 0 does not come before argument 1");
```
```suggestion
                        b1.printf("Argument 0 does not come before argument 1%n");
```

Ladicek commented:
Additions LGTM. Great examples BTW!

gsmet commented:
FWIW even if incomplete, I think it would be worth merging to make it more accessible.

dmlloyd commented:
Fixed the aformentioned problems, plus some bad import formatting, and added:

* Logical and/or
* Switch
* gotoCase/gotoDefault
* loop

I'd ask that any further edits, fixes, and additions be done as a separate PR.
The manual is about 90% done but I cannot work on it further for a few weeks.

Thanks for all the reviews!

gsmet commented:
> It is said right in the first sentence, but it's small and I think a lot of people will miss it.

We need MOOOORE examples!
